src/main/flight/pid.c

```diff
@@ -292,13 +295,127 @@ static void pidApplyRateController(const pidProfile_t *pidProfile, pidState_t *p
+    if (navHeadingState != NAV_HEADING_CONTROL_NONE) {
+        // Apply maghold only if heading control is in auto mode
+        if (navHeadingState == NAV_HEADING_CONTROL_AUTO) {
+            return MAG_HOLD_ENABLED;
+        }
+    }
+    else
+#endif
+    if (ABS(rcCommand[YAW]) < 15 && FLIGHT_MODE(MAG_MODE)) {
+        return MAG_HOLD_ENABLED;
+    } else {
+        return MAG_HOLD_UPDATE_HEADING;
+    }
+
+    return MAG_HOLD_UPDATE_HEADING;
+}
+
+/*
+ * MAG_HOLD P Controller returns desired rotation rate in dps to be fed to Rate controller
+ */
+float pidMagHold(const pidProfile_t *pidProfile)
+{
+
+    static filterStatePt1_t magHoldRateFilter;
+    float magHoldRate;
+
+    int16_t error = DECIDEGREES_TO_DEGREES(attitude.values.yaw) - magHoldTargetHeading;
+
+    /*
+     * Convert absolute error into relative to current heading
+     */
+    if (error <= -180) {
+        error += 360;
+    }
+
+    if (error >= +180) {
+        error -= 360;
+    }
+
+    /*
+        New MAG_HOLD controller work slightly different that previous one.
+        Old one mapped error to rotation speed in following way:
+            - on rate 0 it gave about 0.5dps for each degree of error
+            - error 0 = rotation speed of 0dps
+            - error 180 = rotation speed of 96 degrees per second
+            - output
+            - that gives about 2 seconds to correct any error, no matter how big. Of course, usually more because of inertia.
+        That was making him quite "soft" for small changes and rapid for big ones that started to appear
+        when iNav introduced real RTH and WAYPOINT that might require rapid heading changes.
+
+        New approach uses modified principle:
+            - manual yaw rate is not used. MAG_HOLD is decoupled from manual input settings
+            - instead, mag_hold_rate_limit is used. It defines max rotation speed in dps that MAG_HOLD controller can require from RateController
+            - computed rotation speed is capped at -mag_hold_rate_limit and mag_hold_rate_limit
+            - Default mag_hold_rate_limit = 40dps and default MAG_HOLD P-gain is 40
+            - With those values, maximum rotation speed will be required from Rate Controller when error is greater that 30 degrees
+            - For smaller error, required rate will be proportional.
+            - It uses LPF filter set at 2Hz to additionally smoothen out any rapid changes
+            - That makes correction of smaller errors stronger, and those of big errors softer
+
+        This make looks as very slow rotation rate, but please remember this is automatic mode.
+        Manual override with YAW input when MAG_HOLD is enabled will still use "manual" rates, not MAG_HOLD rates.
+        Highest possible correction is 180 degrees and it will take more less 4.5 seconds. It is even more than sufficient
+        to run RTH or WAYPOINT missions. My favourite rate range here is 20dps - 30dps that gives nice and smooth turns.
```

Comment:
Just noticed that you did some proper commenting, awesome!
I don't think there is a right or wrong answer on rate here, IMHO it's a matter of size of copter and taste!
For example, taking 4.5 seconds to turn around might be right for a big quad, but on my 250 racer that would just looks silly! After tuning i was happy at around 180dps or 1 seconds.

Perhaps it would be a good idea to default mag_hold_rate_limit to half of yaw_rate? That way no nav-specific tuning would be required. Your quad would have slow and gentle yaw, and my racer... would still be a racer! Thoughts?